flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/sink/AvroGenericRecordToRowDataMapper.java

```diff
@@ -41,13 +45,114 @@
 public class AvroGenericRecordToRowDataMapper implements MapFunction<GenericRecord, RowData> {
 
   private final AvroToRowDataConverters.AvroToRowDataConverter converter;
+  private final Schema avroSchema;
+  private final Function<GenericRecord, RowData> customConverter;
 
-  AvroGenericRecordToRowDataMapper(RowType rowType) {
+  AvroGenericRecordToRowDataMapper(RowType rowType, Schema avroSchema) {
+    this.avroSchema = avroSchema;
     this.converter = AvroToRowDataConverters.createRowConverter(rowType);
+    // Check if we need custom conversion for timestamp-nanos
+    this.customConverter = needsCustomConversion(avroSchema) ? this::convertWithNanos : null;
+  }
+
+  private boolean needsCustomConversion(Schema schema) {
+    for (Schema.Field field : schema.getFields()) {
+      if (field.schema().getLogicalType() instanceof LogicalTypes.TimestampNanos) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private RowData convertWithNanos(GenericRecord genericRecord) {
+    RowData baseRowData = (RowData) converter.convert(genericRecord);
+    GenericRowData result = new GenericRowData(baseRowData.getArity());
```

Comment:
I don't really like this solution. We creating 2 `RowData` objects for a single conversion.
Can we do this with a single conversion?